storage/testonly/admin_storage_tester.go

```diff
@@ -126,25 +127,199 @@ func (tester *AdminStorageTester) TestCreateTree(t *testing.T) {
-		if err != nil {
-			t.Errorf(":%v: GetTree() = (%v, %v), want = (%v, nil)", i, newTree.TreeId, err, newTree)
+			t.Errorf(":%v: getTree() = (%v, %v), want = (%v, nil)", i, newTree.TreeId, err, newTree)
 			continue
 		}
 		wantTree = *storedTree
 		if !reflect.DeepEqual(newTree, &wantTree) {
 			t.Errorf("%v: newTree = \n%v, wantTree = \n%v", i, newTree, &wantTree)
 		}
-		if err := readOnlyTX.Commit(); err != nil {
-			t.Errorf("%v: Commit() = %v, want = nil", i, err)
+	}
+}
+
+// TestUpdateTree tests AdminStorage Tree updates.
+func (tester *AdminStorageTester) TestUpdateTree(t *testing.T) {
+	s := tester.NewAdminStorage()
+
+	unrelatedTree, err := createTree(s, MapTree)
+	if err != nil {
+		t.Fatalf("createTree() = (_, %v), want = (_, nil)", err)
+	}
+
+	referenceLog := *LogTree
+
+	validLog := referenceLog
+	validLog.TreeState = trillian.TreeState_FROZEN
+	validLog.DisplayName = "Frozen Tree"
+	validLog.Description = "A Frozen Tree"
+	validLogFunc := func(t *trillian.Tree) {
+		t.TreeState = validLog.TreeState
+		t.DisplayName = validLog.DisplayName
+		t.Description = validLog.Description
+	}
+
+	invalidLogFunc := func(t *trillian.Tree) {
+		t.TreeState = trillian.TreeState_UNKNOWN_TREE_STATE
+	}
+
+	readonlyChangedFunc := func(t *trillian.Tree) {
+		t.TreeType = trillian.TreeType_MAP
+	}
+
+	referenceMap := *MapTree
+	validMap := referenceMap
+	validMap.DisplayName = "Updated Map"
+	validMapFunc := func(t *trillian.Tree) {
+		t.DisplayName = validMap.DisplayName
+	}
+
+	// Test for an unknown tree outside the loop: it makes the test logic simpler
+	if _, errOnUpdate, err := updateTree(s, -1, func(t *trillian.Tree) {}); err == nil || !errOnUpdate {
+		t.Errorf("updateTree() = (_, %v, %v), wanted lookup error (ie, error on update)", errOnUpdate, err)
+	}
+
+	tests := []struct {
+		desc                 string
+		createTree, wantTree *trillian.Tree
+		updateFunc           func(*trillian.Tree)
+		wantErr              bool
+	}{
+		{
+			desc:       "validLog",
+			createTree: &referenceLog,
+			updateFunc: validLogFunc,
+			wantTree:   &validLog,
+		},
+		{
+			desc:       "invalidLog",
+			createTree: &referenceLog,
+			updateFunc: invalidLogFunc,
+			wantErr:    true,
+		},
+		{
+			desc:       "readonlyChanged",
+			createTree: &referenceLog,
+			updateFunc: readonlyChangedFunc,
+			wantErr:    true,
+		},
+		{
+			desc:       "validMap",
+			createTree: &referenceMap,
+			updateFunc: validMapFunc,
+			wantTree:   &validMap,
+		},
+	}
+	for _, test := range tests {
+		createdTree, err := createTree(s, test.createTree)
+		if err != nil {
+			t.Errorf("createTree() = (_, %v), want = (_, nil)", err)
+			continue
+		}
+
+		updatedTree, errOnUpdate, err := updateTree(s, createdTree.TreeId, test.updateFunc)
+		if err != nil && !errOnUpdate {
+			t.Errorf("%v: updateTree() failed with non-Update error: %v", test.desc, err)
+			continue
+		}
+
+		if hasErr := err != nil; hasErr != test.wantErr {
```

Comment:
Personal preference, but I find it easier to read tests that are structured like:
```go
if test.wantErr {
  if err == nil {
    t.Errorf("%v: updateTree()=(_,_,nil); want (_,_,error)", test.desc)
  }
  continue
}
if err != nil {
  t.Errorf("%v: updateTree()=(_,_,%v); want (_,_,nil)", test.desc, err)
  continue
}
```